bookkeeper-server/src/main/java/org/apache/bookkeeper/client/EnsembleUtils.java

```diff
@@ -0,0 +1,98 @@
+ *
+ */
+package org.apache.bookkeeper.client;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.bookkeeper.net.BookieSocketAddress;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class EnsembleUtils {
+    private static final Logger LOG = LoggerFactory.getLogger(EnsembleUtils.class);
+
+    static List<BookieSocketAddress> replaceBookiesInEnsemble(BookieWatcher bookieWatcher,
+                                                              long ledgerId,
+                                                              LedgerMetadata metadata,
+                                                              List<BookieSocketAddress> oldEnsemble,
+                                                              Map<Integer, BookieSocketAddress> failedBookies)
+            throws BKException.BKNotEnoughBookiesException {
+        List<BookieSocketAddress> newEnsemble = new ArrayList<>(oldEnsemble);
+
+        int ensembleSize = metadata.getEnsembleSize();
+        int writeQ = metadata.getWriteQuorumSize();
+        int ackQ = metadata.getAckQuorumSize();
+        Map<String, byte[]> customMetadata = metadata.getCustomMetadata();
+
+        Set<BookieSocketAddress> exclude = new HashSet<>(failedBookies.values());
+
+        int replaced = 0;
+        for (Map.Entry<Integer, BookieSocketAddress> entry : failedBookies.entrySet()) {
+            int idx = entry.getKey();
+            BookieSocketAddress addr = entry.getValue();
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("[EnsembleChange-L{}] replacing bookie: {} index: {}", ledgerId, addr, idx);
+            }
+
+            if (!newEnsemble.get(idx).equals(addr)) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("[EnsembleChange-L{}] Not changing failed bookie {} at index {}, already changed to {}",
+                              ledgerId, addr, idx, newEnsemble.get(idx));
+                }
+                continue;
+            }
+            try {
+                BookieSocketAddress newBookie = bookieWatcher.replaceBookie(
+                        ensembleSize, writeQ, ackQ, customMetadata, newEnsemble, idx, exclude);
+                newEnsemble.set(idx, newBookie);
+
+                replaced++;
+            } catch (BKException.BKNotEnoughBookiesException e) {
+                // if there is no bookie replaced, we throw not enough bookie exception
+                if (replaced <= 0) {
+                    throw e;
+                } else {
+                    break;
+                }
+            }
+        }
+        return newEnsemble;
+    }
+
+    static Set<Integer> diffEnsemble(List<BookieSocketAddress> e1,
+                                     List<BookieSocketAddress> e2) {
+        checkArgument(e1.size() == e2.size(), "Ensembles must be of same size");
```

Comment:
How is this better than Assert?

Reply:
asserts get turned off by default at runtime.